url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/parser/Parser.java

```diff
@@ -0,0 +1,1216 @@
+package de.codesourcery.dcpu16.parser;
+
+import java.io.BufferedReader;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+
+import de.codesourcery.dcpu16.ast.AST;
+import de.codesourcery.dcpu16.ast.ASTNode;
+import de.codesourcery.dcpu16.ast.ArrayInitializer;
+import de.codesourcery.dcpu16.ast.Block;
+import de.codesourcery.dcpu16.ast.DoLoopNode;
+import de.codesourcery.dcpu16.ast.ForLoopNode;
+import de.codesourcery.dcpu16.ast.FunctionDeclarationNode;
+import de.codesourcery.dcpu16.ast.FunctionDefinitionNode;
+import de.codesourcery.dcpu16.ast.FunctionInvocation;
+import de.codesourcery.dcpu16.ast.IfNode;
+import de.codesourcery.dcpu16.ast.InlineAssemblyNode;
+import de.codesourcery.dcpu16.ast.NumberLiteralNode;
+import de.codesourcery.dcpu16.ast.OperatorNode;
+import de.codesourcery.dcpu16.ast.ParameterDeclaration;
+import de.codesourcery.dcpu16.ast.ReturnNode;
+import de.codesourcery.dcpu16.ast.StatementNode;
+import de.codesourcery.dcpu16.ast.StringLiteralNode;
+import de.codesourcery.dcpu16.ast.TermNode;
+import de.codesourcery.dcpu16.ast.VariableDefinition;
+import de.codesourcery.dcpu16.ast.VariableReferenceNode;
+import de.codesourcery.dcpu16.ast.WhileLoopNode;
+import de.codesourcery.dcpu16.compiler.DataType;
+import de.codesourcery.dcpu16.compiler.IScope;
+import de.codesourcery.dcpu16.compiler.Identifier;
+import de.codesourcery.dcpu16.compiler.Operator;
+import de.codesourcery.dcpu16.compiler.Scope;
+import de.codesourcery.dcpu16.exceptions.ParseException;
+import de.codesourcery.dcpu16.parser.ExpressionToken.ExpressionTokenType;
+import de.codesourcery.dcpu16.parser.ShuntingYard.MismatchedClosingParensException;
+import de.codesourcery.dcpu16.util.IInputStreamProvider;
+import de.codesourcery.dcpu16.util.ITextRegion;
+import de.codesourcery.dcpu16.util.TextRegion;
+
+public class Parser
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This class references 47 other classes. By comparison, 98% of the classes in the CodeGuru reference dataset reference fewer. This indicates that this class is highly coupled with other classes. A class that is highly coupled with other classes is difficult to understand and its behavior might change unexpectedly when one of its referenced classes is updated. High coupling could also increase the integration test complexity, maintenance cost and technical debt. We recommend that you simplify this class or break it into multiple classes.